core/src/compression/stream.cc

```diff
@@ -27,6 +28,111 @@
 using namespace com::centreon::broker;
 using namespace com::centreon::broker::compression;
 
+/**
+ * Compression function
+ *
+ * @param data the data to compress.
+ * @param compression_level The compression level, by default -1.
+ *
+ * @return The same data compressed.
+ */
+static QByteArray _compress(QByteArray const& data, int compression_level) {
+  if (data.isEmpty())
+    return QByteArray(4, '\0');
+
+  int nbytes = data.size();
+
+  if (compression_level < -1 || compression_level > 9)
+    compression_level = -1;
+
+  unsigned long len = nbytes + nbytes / 100 + 13;
+  QByteArray retval;
+  int res;
+  do {
+    retval.resize(len + 4);
+    res = ::compress2(
+        (uchar*)retval.data() + 4,
+        &len,
+        (uchar*)data.constData(),
+        nbytes,
+        compression_level);
+
+    switch (res) {
+      case Z_OK:
+        retval.resize(len + 4);
+        retval[0] = (nbytes >> 24) & 0xff;
+        retval[1] = (nbytes >> 16) & 0xff;
+        retval[2] = (nbytes >> 8) & 0xff;
+        retval[3] = (nbytes & 0xff);
+        break;
+      case Z_MEM_ERROR:
+        logging::error(logging::medium)
+          << "compression : Z_MEM_ERROR: Not enough memory";
+        retval.resize(0);
+        break;
+      case Z_BUF_ERROR:
+        len <<= 1;
+        break;
+    }
+  } while (res == Z_BUF_ERROR);
+
+  return retval;
+}
+
+/**
+ * Uncompress function
+ *
+ * @param data The data to extract.
+ * @param nbytes The data size in bytes.
+ *
+ * @return the extract data
+ */
+static QByteArray _uncompress(const uchar* data, int nbytes) {
+  if (!data) {
+    logging::info(logging::medium)
+      << "compression: uncompress - Data is null";
+    return QByteArray();
+  }
+  if (nbytes <= 4) {
+    if (nbytes < 4 || (data[0] != 0 || data[1] != 0
+                   || data[2] != 0 || data[3] != 0))
+      logging::error(logging::medium)
+        << "compression: uncompress - Input data is corrupted";
+    return QByteArray();
+  }
+  ulong expected_size = (data[0] << 24) | (data[1] << 16)
+                     | (data[2] <<  8) | data[3];
+  ulong len = qMax(expected_size, 1ul);
+  QByteArray uncompressed_array(len, '\0');
+
+  forever {
```

Comment:
Qt style. Please remove the loop as we have the expected size. Any kind of error from uncompress() can be processed immediately.

Z_OK = return buffer as is.
Z_MEM_ERROR = throw out of memory exception (std::bad_alloc or exceptions::msg).
Z_BUF_ERROR || Z_DATA_ERROR = data corruption error. Let caller process it.